poptorch_experimental_addons/_impl/core.py

```diff
@@ -102,12 +102,174 @@ def distance_matrix(tensor1: Tensor, tensor2: Tensor, p: int) -> Tensor:
+
+    rounding -- either "nearest" or "stochastic"
+
+    fwd -- whether to quantise the forward value
+
+    bwd -- whether to generate & whether to quantise the gradient:
+           bwd=None  -- no gradient
+           bwd=False -- unquantised gradient (straight-through estimator)
+           bwd=True  -- quantised gradient
+    """
+    if rounding not in ["nearest", "stochastic"]:
+        raise ValueError(
+            "Expected quantise(rounding=?) to be 'nearest' or 'stochastic'"
+            f", actual '{rounding}'"
+        )
+
+    if poptorch.isRunningOnIpu():
+        max_exponent_bits = 5
+        max_mantissa_bits = 10
+    else:
+        max_exponent_bits = 8
+        max_mantissa_bits = 23
+    if exponent_bits > max_exponent_bits:
+        raise ValueError(
+            f"quantise_fpx(exponent_bits={exponent_bits}) not supported, maximum"
+            f" number of exponent bits is {max_exponent_bits}"
+        )
+    if mantissa_bits > max_mantissa_bits:
+        raise ValueError(
+            f"quantise_fpx(mantissa_bits={mantissa_bits}) not supported, maximum"
+            f" number of mantissa bits is {max_mantissa_bits}"
+        )
+
+    q: Tensor
+    if poptorch.isRunningOnIpu():
+        (q,) = poptorch.custom_op(
+            name="SimulatedQuant",
+            domain_version=1,
+            domain="ai.graphcore",
+            inputs=[x],
+            example_outputs=[x],
+            attributes=dict(
+                root_path=str(Path(__file__).parent.parent.absolute()),
+                exponent_bits=exponent_bits,
+                mantissa_bits=mantissa_bits,
+                rounding=rounding,
+                fwd=fwd,
+                bwd={True: "quantise", False: "ste", None: "stop"}[bwd],
+            ),
+        )
+        return q
+
+    def _quantise(x: Tensor) -> Tensor:
+        max_exponent = 2 ** (exponent_bits - 1) - 1
+        absmax = 2**max_exponent * (2 - 2**-mantissa_bits)
+        downscale = 2.0 ** (126 - max_exponent)
+        mask = torch.tensor(
+            2 ** (23 - mantissa_bits) - 1, dtype=torch.int32, device=x.device
+        )
+        offset = (
+            torch.randint(  # type:ignore[call-overload]
+                0, mask + 1, x.shape, dtype=torch.int32, device=x.device
+            )
+            if rounding == "stochastic"
+            else mask // 2
+        )
+        # Manually clip to max
+        # Then scale down (to generate appropriate subnormals) & mask off mantissa bits
+        q = x.to(torch.float32)
+        q = torch.clip(x, -absmax, absmax)
+        q /= downscale
+        q = ((q.view(torch.int32) + offset) & ~mask).view(torch.float32)
+        q *= downscale
+        q = q.to(x.dtype)
+        return q
+
+    class F(torch.autograd.Function):
+        @staticmethod
+        def forward(  # type:ignore[override]
+            ctx: torch.autograd.function.FunctionCtx, xx: Tensor
+        ) -> Tensor:
+            return _quantise(xx) if fwd else xx.clone()
```

Comment:
I feel like I've come across this before, but is there a reason you need to clone in fwd but not bwd?

Reply:
Ah, it's a bit of a mystery, I recall it was quite a complex error message when I didn't `clone()` here. Perhaps double-differentiation would require a `clone()` in bwd too(?)